src/compiler/crystal/semantic/restrictions.cr

```diff
@@ -1175,6 +1175,15 @@ module Crystal
       end
 
       other.type_vars.each_with_index do |other_type_var, i|
+        # If checking the return type, any type matches Nil
+        if i == other.type_vars.size - 1 && nil_type?(other_type_var, context)
+          # Also, all other types matched, so the matching type is this proc type
+          # except that it has a Nil return type
```

Comment:
I thought the proc were honoring subtyping of arrow types.

It is safe for the following example to compile.

```cr
class Foo
end
 
class Bar < Foo
end
 
def foo(x : Proc(Bar, Nil))
  x
end
 
a = 1
 
proc = foo(->(b : Foo) { a = 2 })
r = proc.call(Bar.new)
```

It is subject for a later change, but is the subtying of arrow type in procs that we are missing it seems.

Reply:
Yeah, we need to make subtyping work, but it won't work all of the time: only for class references. For structs it's not possible because the representation is different. Also for unions, I think. So for now I didn't implement it until we think it well. But matching against `Nil` is more common and easier to check.